### PR TITLE
security: fix REDOS vulnerabilities

### PR DESCRIPTION
**Pull request by davisjam**
Demonstrate and fix catastrophic backtracking in two regexes.

For #1070.

### CODE REVIEW

**Comment by davisjam** [comment]
Uhm. The tests will presumably not pass here because they are demonstrating a DoS. Hope there's a timeout on Travis.

**Comment by davisjam** [comment]
Aww poor Travis.

"Ran for 22 min 45 sec
Total time 45 min 9 sec"

**Comment by joshbruce** [comment]
@davisjam: Thank you so much for the contribution. Nah, now we know the test works. 

I can confirm it crapped out on the DOS test; so, I'm gonna call that good enough evidence that we have the right test.

Not sure what's up with that since the solution should be in the PR itself though?? Maybe something needs to change with the test (tagging @UziTech and gonna assign put up a review notice for the rest of the committers).

screen shot 2018-02-26 at 2 16 12 pm

**Comment by davisjam** [comment]
@joshbruce I fixed the html (that's test # 52 redos-html-closing) but not the nolink rule. So one of the attacks still works.

**Comment by joshbruce** [comment]
@davisjam: I may not be following. Ah! I think I see.

Yes, 52 completed successfully; so, that's good (I don't think our tests log when they have started - only when they succeed or fail). Therefore, my hypothesis is that #53 is the test/new/redos-nolink.html one, which I missed somehow - thought there was only one test.

Maybe rename the test A-redos-nolink.html - the tests seem to run in alphabetical order?? (Seems like the cheapest experiment to test this hypothesis given our current setup.)

**Comment by davisjam** [comment]
No need for a hypothesis -- if I remove the nolink test, all tests pass on my fork.

66bc339 adds a time check -- tests that take more than 1 second will fail now.

**Comment by joshbruce** [comment]
@davisjam: Interesting solution, I like it. Definitely gets us moving in the right direction on performance testing. We should be able to adjust this up and down, yeah? (Not getting too sidetracked here, just want to make sure I have the information for adding to #963 and incorporating conversations @UziTech, @Feder1co5oave, and I have had re benchmarking.)

Travis CI is still hung on 53. Thinking the timer solution isn't working because the function never returns. It's almost like we need to have the succeed-fail timer be an asynchronous thing??

**Comment by joshbruce** [comment]
Specifically the `testFile(engine, file, filename, i + 1);` function.

**Comment by davisjam** [comment]
@joshbruce Right, 66bc339 won't abort long-running tests. It just adds another test for correctness -- tests shouldn't take more than one second.

To abort long-running tests stuck on REDOS, you would need to either (1) run in a Node VM (it's a core Node module), or (2) run in a child process.

**Comment by joshbruce** [comment]
@davisjam: Yeah, would probably want to go with the child process.

So, how do we validate that test 54 is the redos-nolink one?? Or, will we just assume it is, at which point this only partially corrects the REDOS issues, yeah??

**Comment by davisjam** [comment]
@joshbruce

1. Tests in child processes: yes, maybe open that as a separate issue.
2. Test ordering: I just pushed 9a11030, which adds a 'print for beginning a test'. Merge if you want. #1084.
3. I've slightly tweaked the language parsed by the noline regex. See notes in my [commit message](https://github.com/markedjs/marked/pull/1083/commits/85e8c36ccba971c2a20efb5c1184ca36d7fc1c87).

Both REDOS issues (actually all 4, but two pairs of related regexes) are fixed by this PR. Again, not sure if I broke anything, so definitely needs review for the nolink pattern.

**Comment by joshbruce** [comment]
@davisjam: We're gonna be breaking all the things before this is resolved. Good thing it's a GitHub repository. 😂 

I can't add this comment in the PR itself...sorry.

`/test/index.js` - line 190: 

```js
// before
console.log('#%d. %s completed.', index, filename);

// after
console.log('completed %s', index, filename);
```

The log is difficult to read to verify the tests actually completed - failed and CI failure is awesome because it blows things up.

The failing test here is the only one allowed. (I'm updating templates to make that obvious.)

screen shot 2018-02-26 at 3 57 10 pm

File: test/index.js [hunk 1]
```diff
@@ -143,6 +148,8 @@ function testFile(engine, file, filename, index) {
     delete marked._original;
   }
 
+  console.log('#%d. Beginning %s.', index, filename);
+
   if (opts.length) {
     marked._original = marked.defaults;
     marked.defaults = {};
```
Comment (joshbruce) [inline test/index.js]: `/test/index.js` - line 190: 

```js
// before
console.log('#%d. %s completed.', index, filename);

// after
console.log('completed %s', index, filename);
```

Comment (davisjam) [reply test/index.js]: Addressed in 3066fd2

**Comment by joshbruce** [comment]
I suppose I could make the change myself, if you're okay with me hopping into your branch directly...??

**Comment by davisjam** [comment]
Changed already in d1a85c6, let me know what you think.

File: test/index.js [hunk 2]
```diff
@@ -187,7 +194,7 @@ function testFile(engine, file, filename, index) {
     }
   }
 
-  console.log('#%d. %s completed.', index, filename);
+  console.log('    #%d passed (%s)', index, filename);
```
Comment (joshbruce) [inline test/index.js]: Like where you're going. Sorry for not being explicit in what the brain-hiccup was.

The double digit is what made it difficult to read for me.

```
52. beginning something something
52. something something completed
```

So getting rid of the digit.

```
52. beginning something something
failed
```

And

```
52. beginning something something
failed
```

Might be easier. Maybe even adding a blank line after each group??

```
52. beginning something something
failed

53. beginning something something
passed
```

**Comment by davisjam** [comment]
@joshbruce Let's not worry about test output formatting for now. You can tackle that later in a separate issue if you like.

I'm more interested in your thoughts (and those of the larger team, though I don't know whom to tag) on the security portion of this PR.

In particular the changes to the regexes. See [here](https://github.com/markedjs/marked/pull/1083/commits/85e8c36ccba971c2a20efb5c1184ca36d7fc1c87).

**Comment by joshbruce** [comment]
@davisjam: Fair enough. We're just _so_ close and my perfectionist/UX side couldn't drop it. 😆 

Thanks for calling me out in a civil way. Will make a ticket.

@Feder1co5oave seems to be the resident regex guru, to be honest. I've never met anyone who can actually regex the way he seems to.

@UziTech is the resident test harness guru, he made some pretty cool upgrades to it.

I think the solution is sound. The tests are there to guard against regress. That plan as I have it in my head right now given timezones and whatnot:

1. Let this PR cool for a minute and allow others to take a look and comment on all the things.
2. Having said that, I'm comfortable taking the chance, given the surrounding context and will take ownership of any fallout from **merging this by 11am EST February 27th (tomorrow for me)**.
3. Will also merge #1081, which should now help a bit with the who to tag for what via the AUTHORS page.
4. Will pull into #1076 (our standing release candidate, if you will).
5. Finish the release checklist and go.

This does break slightly with some of the protocols and process but I think the PRs have been around long enough for review and we can always change things later, if need be.

**Comment by joshbruce** [comment]
I would rather create a release that definitively fixes a security vulnerability that has tests against regression into the vulnerable posture and then have to take the heat for accidentally breaking something else than to have Marked go back onto the blacklists so soon after its removal.

**Comment by davisjam** [comment]
@joshbruce Sounds good. I appreciate your security-first stance.

I removed the test pretty printing from this PR. It's now in #1085.

**Comment by joshbruce** [comment]
@davisjam: Thanks. And thank for the wonderful contribution and insight all around, really appreciate it. 

Do you mind if me and some of the crew update the branch directly from your fork re #1085??

**Comment by davisjam** [comment]
@joshbruce Not sure why you would want to combine #1083 (this PR, security patch) and #1085 (prettifying test behavior)?

**Comment by joshbruce** [comment]
@davisjam: Sorry. I'm gonna blame text-based communication and possible ignorance. I'm not accustomed to having distributed collaborators; or a lot of them, to be honest.

Let's take this to #1085, as it has nothing to do with #1083. :)

**Comment by joshbruce** [comment]
I'll fix the conflicts...

**Comment by davisjam** [comment]
@joshbruce Life is suffering. Thanks for fixing.

**Comment by joshbruce** [comment]
@davisjam: Ah, Buddhism. I watched an interview once where a monk said, "It's not a very accurate translation. It more closely translates to, 'life is filled with disappointment'" - which makes the whole "no expectations" make so much more sense. 😊 

No worries. Totally my "fault" for merging in the changes to `/test/index.js` before this one. Only right that I would fix the mess I made. 😉

**Comment by davisjam** [comment]
@joshbruce I find the mis-translation a good rule of thumb for software development, anyway ;-)

**Comment by joshbruce** [comment]
@davisjam: lol - indeed. On that note, check #1087 just for future reference. We're getting there. Does your editor recognize `.editorconfig` settings and rules??

**Comment by davisjam** [comment]
> Does your editor recognize .editorconfig settings and rules??

I use vanilla vim. So no.

**Comment by joshbruce** [comment]
@davisjam: Cool. No worries. Just be sure to run `npm run lint` before submitting PRs in the future and the tabs vs. spaces debate will be fixed for us. 😉

**Comment by davisjam** [comment]
Ahha, another opportunity for an issue.

The README is [silent](https://github.com/markedjs/marked#running-tests--contributing) on this topic.

Personally I like it when the 'test' task incorporates the linter.

**Comment by joshbruce** [comment]
@davisjam: Ask and you shall receive. See also #1081 (updating all the documentation - still working on style - but, like I said, we're getting there) and #1076 (updating the PR and Issue templates).

File: test/new/redos-html-closing.md [hunk 3]
```diff
@@ -0,0 +1 @@
+<tag		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""		""<"	/>a'a
```
Comment (Feder1co5oave) [inline test/new/redos-html-closing.md]: can we use spaces instead of tabs?
why is redos-html-closing.html empty?
can you use underscores instead of dashes in the test name to conform to other tests?

File: test/new/redos-nolink.md [hunk 4]
```diff
@@ -0,0 +1 @@
+![\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]\[[]!\
```
Comment (Feder1co5oave) [inline test/new/redos-nolink.md]: why is redos-newlink.html empty?

Comment (joshbruce) [reply test/new/redos-nolink.md]: What would be expected otherwise? Think in this case we're not so much worried that the write HTML gets spit out and more concerned on whether the test can finish - given that Travis died...literally on it, I'd say it's an effective test. :)

File: test/new/redos-html-closing.md (hunk 3 above)
Comment (joshbruce) [reply test/new/redos-html-closing.md]: 1. Can we create an issue to examine that question? (Not sure if they absolutely need to be tabs or not, would like to know the answer.)
2. See other comment.
3. Recommend fixing that in a different PR? (@davisjam contributed a lot today and it seems like relatively small things compared to the bigger problem. He called me out on the same thing when I was trying to make the testing harness prettier - and even did follow up on that.)

These don't seem like dealbreakers for merging to me right now.

Comment (joshbruce) [reply test/new/redos-html-closing.md]: ps. @UziTech and I made the testing harness prettier at the same time and those changes have already been merged.

File: test/new/redos-nolink.md (hunk 4 above)
Comment (davisjam) [reply test/new/redos-nolink.md]: I figured that was to hold the expected output. I don't know what the expected output is, just so long as it gets generated without REDOS.

File: test/new/redos-html-closing.md (hunk 3 above)
Comment (davisjam) [reply test/new/redos-html-closing.md]: > can we use spaces instead of tabs?

Why?

>  why is redos-html-closing.html empty?

See below; I don't know what the expected output is.

>  can you use underscores instead of dashes in the test name to conform to other tests?

Yes, will fix.

**Comment by davisjam** [comment]
I rebased on master and made the changes requested by @Feder1co5oave.

**Comment by joshbruce** [comment]
Wow! Thank you so much.

All right. Gonna stick to the plan re by 11am and say the following.

1. @davisjam, you want to be a committer? I think it would be nice to have a person who really knows security, is very responsive and easy going, and (at least so far) seems to be a benevolent soul able to help. No need to answer right away.
2. We haven't fleshed out [governance and process](https://opensource.guide/leadership-and-governance/) on the community side as much as I think we will get to later. Having said that, we've been operating under the BDFL model for a long time and I think we're still in that mode. Further, while I appreciate the notion a great deal, I think it can actually be taken a step farther and I bring the [8fold values, principles, and practices](https://8fold.pro/about) wherever I go; therefore, would like to try and move us toward something more like the [advice process](http://www.reinventingorganizationswiki.com/Decision_Making).
3. We haven't had this many committers in a long time - and in the old world none of us were it; so, it might take some getting used to. As such, I've updated #1076 to reflect the language of the AUTHORS regarding who's filling out which parts of the checklists. So, I'm not sure everyone knows what is okay and what's not at the moment (everyone might be waiting for me to merge this or waiting to tell me to).
4. Thank you so much! Seriously. This is _huge_ for us at this stage in reviving Marked in ways that go beyond fixing the security vulnerability.

**Comment by joshbruce** [comment]
ps. "We" being the Marked project itself - not the four of us on this latest active crew.

**Comment by Feder1co5oave** [comment]
@joshbruce I'm getting a little annoyed at you merging essential code changes without my review.

**Comment by Feder1co5oave** [comment]
Text fixtures in this PR are wrong. They only pass because the comparing algorithm is broken.

**Comment by davisjam** [comment]
Yes, I know the html files don't match the md. I didn't know what to put for the expected content. Since the test suite didn't complain I did not pursue it further.

Sounds like a problem in the test harness? Is there an issue for this?

**Comment by UziTech** [comment]
I agree. We really need to slow things down and get them right.

@davisjam I think #1017 should fix it.

**Comment by joshbruce** [comment]
@Feder1co5oave and all: I can see that. Having said that, I stated what the plan was. There was no pushback regarding that plan. 

The comparing algorithm being *wrong* did not seem to be part of the review (unless we're referring to the lack of HTML output, of which there was not an alternative presented) - only that the file names did not follow convention, tabs vs spaces, and so on. The evidence presented by @davisjam seemed sound, thorough, and I watched the DOS happen on Travis. 

If the comparing algorithm is wrong, let's open an issue and discuss again.

Further, when @karenyavine from snyk.io reached out she said, "I'll wait a bit for publishing a publish vulnerability advisory, do you think there is an ETA for this?" (see [comment](https://github.com/markedjs/marked/issues/1070#issuecomment-368544463)). When I asked what the normal protocol was, she said, "our protocol says that if there is a public issue, we can add it." (The time being given was most likely based on the project just now coming back to life and her not wanting to discourage us from continuing to try.) I thanked her for the consideration, explain that sticking to normal snyk.io protocol is fine (nature of open source software development), and did ask for at least 24 hours to attempt a solution before submitting us back to snyk.io - she was kind enough to do so...that's about the time @davisjam stepped in, to give you an idea of how close we are to being back on the list.

To make marked an inviting community for contributors that is sustained and owned by that community it cannot be a one-man show...not even if that "one-man" is me. As I have said elsewhere, I was asked to be here and to take the project on - right now, we don't have a public way of changing that.

I took the project on via the channels and methods afforded to us. We're all doing the best we can to make marked awesome with the knowledge and skills we have. (See also [CODE_OF_CONDUCT PR](https://github.com/markedjs/marked/pull/1094)). 

Been catching a lot of heat lately, which is admittedly part of my job as a [publisher](https://github.com/markedjs/marked/blob/master/AUTHORS.md) but seriously, how did we go from commenting on the awesomeness of working with one another to [storming mode](https://en.wikipedia.org/wiki/Tuckman%27s_stages_of_group_development)? (The code of conduct should help us establish explicit norms though - I do appreciate the NodeJS one linked from the GitHub docs.)

**Comment by davisjam** [comment]
@UziTech 
> I think #1017 should fix it.

Sounds good. I wonder if it will turn up other test failures, too?

Anyway, once #1017 is in, fixing the test failures for the test cases I added will be straightforward -- just copy/paste the output from marked into the html file. For these test cases, "any output is good output". The goal is just for the input not to hang things.

> I agree. We really need to slow things down and get them right.

I agree that "move fast and break things" isn't the best philosophy for a major library (1M+ downloads/month). In @joshbruce's defense, my PR did not introduce regressions as defined by the test suite, and my tweaks were minimal enough that I think any regressions will be corner cases. Since the issue was already publicly disclosed, I think getting the fixes in place was overall a good move.

However, the past is the past, and a code review of the changes is still in order. @Feder1co5oave, the change in the source code was exclusively to regexes. I believe one change was innocuous (new regex matches same language), but I'd urge a review of the second change, described in the [commit message](https://github.com/markedjs/marked/pull/1083/commits/b15e42b67cec9ded8505e9d68bb8741ad7a9590d).

@joshbruce 
> our protocol says that if there is a public issue, we can add it

I'll push back a little on this stance -- I know you're new to this so I mean this constructively, not critically, I promise! Anyway, with the potential security implications of these issues, handling them privately (i.e. not announcing the issue until there's a fix) would have permitted more room to be sure the fix didn't introduce regressions. Open source means the code is open, but it doesn't require that everything happen publicly. In case you don't know, "critical infrastructure"-type projects will often wait until a fix is ready before announcing that there was a vulnerability with a public issue, though of course if they delay too long this becomes problematic.

**Comment by davisjam** [comment]
> Maybe publicly stating what our protocols are under certain extreme circumstances would help curb that for me?

Always good to get this stuff written down, especially for folks like me who don't always know what the preferences are. Something along the lines of "Please disclose potential security issues by email to the project maintainers. We will provide an initial assessment of security reports within 48 hours and should apply patches within 2 weeks."

> Still concerned about potential anti-patterns that can result from "working in the dark" on certain things.

Right, don't want to sit on a vulnerability for year(s) of course.

**Comment by Feder1co5oave** [comment]
> Yes, I know the html files don't match the md. I didn't know what to put for the expected content. Since the test suite didn't complain I did not pursue it further.
> Sounds like a problem in the test harness? Is there an issue for this?

@davisjam don't worry, I really appreciated you stepping in to help us in dealing with these security issues and this contribution was really important. I was looking forward to investigate the tools you suggested.
Those empty html fixture are very easily solvable with
```
bin/marked test/new/redos_html_closing.md > test/new/redos_html_closing.html
bin/marked test/new/redos_nolink.md > test/new/redos_nolink.html
```
That the test suite makes those tests pass is our fault, not yours. I explained why in #1017:
> the current text comparison algorithm is broken: whenever the expected html is shorter than the generated one, and matches everything up until its end, the test is marked as complete, even though marked outputted something more.

edit: I see that you already figured that out. Anyway being that #1017 uses an external library (which is not well maintained) it might bring some issues on its own in the future.

I wanted to check the change you made to the html rule, not because of the change itself, but because the original rule was "naive" to say the least. The nolink one never bothered me.

I realize #1058 was critical and a fix should have been released as soon as possible. I was working on it, but also trying to comply more with commonmark on raw html, so we wouldn't have to do the job twice.
I'm pretty sure although that other regexes need to be checked for backtracking vulnerabilities; I suggest you have a look at
- `block._title`
- `block.table.nptable`
- `block.table.table`
- `inline._inside`
- `inline.gfm._backpedal`

because at a rapid inspection those are the ones with a [star height] > 1.

[star height]: https://en.wikipedia.org/wiki/Star_height

---

@joshbruce I admittedly missed that comment from karen. At one point I had to turn off email notifications: I received a total of **260** of them since #1058 had been opened (feb 18), until feb 26. I really couldn't follow everything that was going on.
Btw that Tuckman thing was a nice read.

**Comment by joshbruce** [comment]
@Feder1co5oave: Hey brother, no worries, we miss things sometimes. It's hard to keep up with conversations sometimes. It's also hard to trust that other people "got this" (or not) when you haven't worked with them for long periods of time. (That lack of face to face though - man it's killing me, not gonna lie.)

Yeah. The Tuckman model is interesting, I'm glad you appreciated it. There's debate over its validity (as with most anything involving human interaction). But, I do keep seeing it over and over again; so, I keep telling people about it. 😊 

I really appreciate game theory as well, specifically cooperative games (I don't do well in building competitive games). It's fascinating how simple rules can lead to some really interesting group dynamics. I will say the "skipping the storming" bit can usually be helped by having a group get together to establish those norms. I facilitated a session like that for the team I'm on for my day job when we first got together...I love that team. I would probably honestly die for that team if it came to that...I'm not kidding or being hyperbolic. And, we are finally getting the influence to do some really cool things - took us three years to get there - but still. 😝 

***

I tend not to assume the level of knowledge in various areas of other people; so, sorry if you know some of this next bit...

On the notification front, I don't know what email client you have, but here are my rules for dealing with all things GitHub (it has saved my skin! - and sanity) - subscriptions across the board actually (this is Mail for macOS):

1. I don't have email notifications turned on - I go to it when I want to, not when it gets all up in my face. See Randy Pausch on [Time Management](https://youtu.be/oTugjssqOT0). "Notifications are interruptions" [and I got stuff to do]. 😉 
2. Here's the rule (if all of the following are true):
  - If "from" contains "@github.com" (all the GitHub emails)
  - If "subject" contains "[account]/[repo]" (all the ones from that repo)
  - If "Message content" does NOT contain "@joshbruce" (not explicitly for me, woohoo!)
  - Then "Move message" to mailbox: "[repo specific mailbox]"
3. I have a day where I do most of my planning and whatnot per week, as part of that, I go through and catch up on all the conversations I might have missed that I may or may not want to comment on. See [The Seven Habits of Highly Effective People](https://www.amazon.com/Habits-Highly-Effective-People-Powerful/dp/0743269519) (specifically "sharpening the saw" - but it's a good read - at least I thought so), [Getting Things Done](https://www.amazon.com/Getting-Things-Done-Stress-Free-Productivity/dp/0143126563/ref=asap_bc?ie=UTF8), and [The Pomodoro Technique](https://francescocirillo.com/pages/pomodoro-technique) (I have the PDF for when this was still being market tested before it became a commercial product, if you would like).
